server/auth/auth.go

```diff
@@ -692,6 +806,51 @@ func (auth *AuthManager) Notify(acctID account.AccountID, msg *msgjson.Message,
 	auth.SendWhenConnected(acctID, msg, timeout, missedFn)
 }
 
+func (auth *AuthManager) registerViolation(user account.AccountID, violation Violation) int32 {
+	auth.violationMtx.Lock()
+	defer auth.violationMtx.Unlock()
+	counts, found := auth.violationCounts[user]
+	if !found {
+		counts = make(violationCounter)
+		auth.violationCounts[user] = counts
+	}
+	counts[violation]++
+	score := counts.score()
+	log.Debugf("Registering outcome %q (badness %d) for user %v, new score = %d",
+		violation.String(), violation.Score(), user, score)
+	return score
+}
+
+// SwapSuccess registers the successful completion of a swap by the given user.
+func (auth *AuthManager) SwapSuccess(user account.AccountID) {
+	auth.registerViolation(user, ViolationSwapSuccess)
+}
+
+// Inaction registers an inaction violation by the user at the given step.
+func (auth *AuthManager) Inaction(user account.AccountID, step NoActionStep) {
+	violation := step.Violation()
+	if violation == ViolationInvalid {
+		log.Errorf("Invalid inaction step %d", step)
+		return
+	}
+	score := auth.registerViolation(user, violation)
+	if score < int32(auth.banScore) {
+		return
+	}
+	log.Debugf("User %v ban score %d is at or above %d. Penalizing.", user, score, auth.banScore)
+	_ = auth.Penalize(user, account.FailureToAct)
+}
+
+// MissedPreimage registers a missed preimage violation by the user.
+func (auth *AuthManager) MissedPreimage(user account.AccountID) {
+	score := auth.registerViolation(user, ViolationPreimageMiss)
+	if score < int32(auth.banScore) {
+		return
+	}
+	log.Debugf("User %v ban score %d is at or above %d. Penalizing.", user, score, auth.banScore)
+	_ = auth.Penalize(user, account.PreimageReveal)
+}
+
 // Penalize signals that a user has broken a rule of community conduct, and that
 // their account should be penalized.
 func (auth *AuthManager) Penalize(user account.AccountID, rule account.Rule) error {
```

Comment:
Correct.  In fact, the `Penalize` method can be unexported now as consumers are just registering violations.

Reply:
Correction - the admin interface is the last place `Penalize` is used outside of the `auth` package now, and that's for a manual ban.

In subsequent violation records work, I expect we would remove the `rule` and even a `suspended` column, unless that's how we choose to handle manual suspension/ban. (It could be done with a synthetic violation with an extremely high score).